### ISSUE TITLE
Gateway API - Connection reset by peer 

### ISSUE THREAD

**Issue by ThomasCardin** [cilium/cilium#30720]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When I make a TLS-enabled HTTPS call to the gateway's FQDN on port 443, I encounter a "connection reset by peer" error. The request appears to reach the cluster IP of the service but is then forcefully terminated.

Request URL:
```bash
https://cilium-gateway-external-gateway.networking.svc.cluster.local:443
```

Service Definition:
```bash
service/cilium-gateway-external-gateway   LoadBalancer   10.110.98.3    192.168.2.100   80:32147/TCP,443:31704/TCP   17m
```

Gateway Configuration (gateway.yaml)
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: external-gateway
  namespace: networking
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
spec:
  gatewayClassName: cilium
  listeners:
  - name: external-gateway-http
    port: 80
    protocol: HTTP
    hostname: "*.{dns}"
    allowedRoutes:
      namespaces:
        from: All
  - name: external-gateway-https
    hostname: "*.{dns}"
    port: 443
    protocol: HTTPS
    allowedRoutes:
      namespaces:
        from: All
    tls: 
      mode: Terminate 
      certificateRefs:
        - name: gateway-tls 
          kind: Secret  
          group: core 
```

HTTPRoute Configuration (httproute.yaml)
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: privatebin-external-nginx
  namespace: privatebin
spec:
  parentRefs:
  - name: external-gateway
    namespace: networking
  hostnames:
    - "privatebin.{dns}"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
    - name: nginx
      port: 80
```

HTTP requests (without TLS)  works. If you need anything at all, please let me know. I would be happy to help :smile: 

### Cilium Version

1.15.0

### Kernel Version

Linux talos-pcn-dn1 6.6.13-talos #1 SMP Wed Jan 24 14:04:57 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux

### Kubernetes Version

v1.29.1

### Relevant log output

```shell
2024-02-12T17:05:37Z ERR Request failed error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: read tcp 10.244.1.174:41380->10.110.98.3:443: read: connection reset by peer" connIndex=2 dest=https://privatebin.{dns}/ event=0 ip=198.41.192.167 type=http
2024-02-12T17:12:12Z ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: read tcp 10.244.1.174:54578->10.110.98.3:443: read: connection reset by peer" cfRay=85467486190739fb-YYZ event=1 ingressRule=2 originService=https://cilium-gateway-external-gateway.networking.svc.cluster.local:443
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for reporting this @ThomasCardin. Could I ask you to explain a bit more about your setup, and upload a sysdump if you can (as per https://docs.cilium.io/en/stable/operations/troubleshooting/#reporting-a-problem)?

I should note that for the Gateway API setup to work, the request has to match the configured DNS name in both the Gateway and the HTTPRoute - so if you're connecting to [https://cilium-gateway-external-gateway.networking.svc.cluster.local:443](https://cilium-gateway-external-gateway.networking.svc.cluster.local/), it won't match `privatebin.{dns}`, even if `{dns}` is set to `cilium-gateway-external-gateway.networking.svc.cluster.local`.

Basically, the Gateway will check the SNI of the HTTPS request to ensure that it matches `privatebin.{dns}`.

**Comment by ThomasCardin** [comment]
Hello, thank you for your response, here is the complete setup:

Gateway API

```yaml
---
# To test this gateway modify the /etc/hosts and set the *.ninebasetwo.net to the LB IP of the external-gateway
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: external-gateway
  namespace: networking
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
spec:
  gatewayClassName: cilium
  listeners:
  - name: external-gateway-http
    port: 80
    protocol: HTTP
    hostname: "*.ninebasetwo.net"
    allowedRoutes:
      namespaces:
        from: All
  - name: external-gateway-https
    hostname: "*.ninebasetwo.net"
    port: 443
    protocol: HTTPS
    allowedRoutes:
      namespaces:
        from: All
    tls: 
      mode: Terminate 
      certificateRefs:
        - name: gateway-tls 
          kind: Secret  
          group: core 
```

The nginx deployment with the used HTTPRoute

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: nginx
spec:
  replicas: 2
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:stable-perl
        ports:
        - containerPort: 80

---
apiVersion: v1
kind: Service 
metadata:
  name: nginx
  namespace: nginx
spec:
  type: ClusterIP
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: http-external-nginx
  namespace: nginx
spec:
  parentRefs:
  - name: external-gateway-https
    namespace: networking
  hostnames:
    - "nginx.ninebasetwo.net"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
    - name: nginx
      port: 80
```

The cloudflared  tunnel configuration (possilby the problem)

```
tunnel: tunnel
    credentials-file: /etc/cloudflared/creds/credentials.json
    metrics: 0.0.0.0:2000

    originRequest:
      http2Origin: true

    no-autoupdate: true
   
    ingress:
    - hostname: nginx.ninebasetwo.net
      service: https://cilium-gateway-external-gateway.networking.svc.cluster.local:443
      originRequest:
        originServerName: ingress.ninebasetwo.net
    - service: http_status:404
```

Regarding the `sysdump`, what information do you need? I see that multiple files are created. On my side, I will continue to investigate the problem taking into account the information you have provided me. Also, note that when I set `http://cilium-gateway-external-gateway.networking.svc.cluster.local:80` for the `service` field inside my cloudflared setup it works correctly.

Thanks!

**Comment by youngnick** [comment]
It seems like cloudflared is running its own tunnel inside the cluster, so because the connection is coming from inside the cluster, you can use the `cluster.local` name, and you don't actually need the Gateway API config. To put this another way, Gateway API is for exposing the service to the internet, but cloudflared doesn't need that because it's already inside the cluster.

**Comment by ThomasCardin** [comment]
Okay, so if I understand correctly, Cloudflared doesn't need the Gateway API since its tunnel running inside the cluster already allows it to expose a svc to the internet? Still, it's odd that the gateway works on port 80 but not on 443 (perhaps there's an issue specifically with Cloudflared's interaction on that port and the cillium gateway api works perfectly fine on 443 without cloudflared). Could you suggest tools to help me investigate this further?

Thanks for your time!

**Comment by youngnick** [comment]
Hmm, because the HTTPS expects the SNI to be set, unless cloudflared has an equivalent config to the `hostname` config to set the SNI, the routing won't work.

When you create the Gateway API config, do requests from a browser from outside the cluster work?

And also, does a request from inside the cluster using something like curl's `--resolve` flag work?

The curl command would look like (from what I can see):
```
curl -vi --resolve ingress.ninebasetwo.net:443:<whatever the cilium-gateway-external-gateway Cluster IP is> https://cilium-gateway-external-gateway.networking.svc.cluster.local:443
```
That will connect to the "outside" of the Gateway API service from inside the cluster, setting the SNI correctly to `ingress.ninebasetwo.net`.

The reason I ask is that we have had some reports (#28254 for example) of hairpinning not working, so I'm trying to narrow this down.

**Comment by ThomasCardin** [comment]
Okay, I've removed the Cloudflared tunnel and configured my system to use a static IP. Now, these commands work as expected (with ports 80 and 443 open on my router) i'm getting a 200:

```bash
curl -vi http://nginx.ninebasetwo.net 
curl -vi https://nginx.ninebasetwo.net 
```

However, when I close port 80 and leave only 443 open, everything stops working. Is it normal? 😅

**Comment by youngnick** [comment]
That does seem weird, but I'm not sure how to help with it. There's no reason why a call to `https` should need access to port 80, and Cilium shouldn't be affected if the traffic is being blocked at your router.

If you could check that the Gateway Service is still serving on 80 and 443 when your router closes the ports, then we can see if it's Cilium that's the problem.

**Comment by ThomasCardin** [comment]
Okay, so if I set the IP of my load balancer to my domain name to test locally via /etc/hosts, I get this error on port 443:

```bash
curl -vi https://nginx.ninebasetwo.net:443
*   Trying 192.168.2.101:443...
* Connected to nginx.ninebasetwo.net (192.168.2.101) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/pki/tls/certs/ca-bundle.crt
*  CApath: none
* Recv failure: Connection reset by peer
* OpenSSL SSL_connect: Connection reset by peer in connection to nginx.ninebasetwo.net:443
* Closing connection 0
curl: (35) Recv failure: Connection reset by peer
```

Maybe I can try with the SNI but now I'm just trying locally, what would be the alternative to ingress.ninebasetwo.net

**Comment by ThomasCardin** [comment]
Maybe a cert manager issue? I did enable this feature ` --feature-gates=ExperimentalGatewayAPISupport=true`

**Comment by ThomasCardin** [comment]
Alright, so I think I'm going to close this issue. Everything seems to be working fine since I've been using a static IP. Thanks for your help! I will reopen it if the problem comes back!

**Comment by arturs706** [comment]
I am getting an error - Waiting for HTTP-01 challenge propagation: wrong status code '404', expected '200'
in the below code
```
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: cilium
spec:
  controllerName: io.cilium/gateway-controller
---
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: tls-gateway
spec:
  gatewayClassName: cilium
  infrastructure:
    annotations:
      io.cilium/lb-ipam-ips: "192.168.3.50"
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    hostname: "*.artdevldn.co.uk"
    allowedRoutes:
      namespaces:
        from: All
  - name: https
    protocol: HTTPS
    port: 443
    hostname: "*.artdevldn.co.uk"
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        name: example-app-tls
    allowedRoutes:
      namespaces:
        from: All
---
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: https-app-route-1
spec:
  parentRefs:
  - name: tls-gateway
  hostnames:
  - "cluster.artdevldn.co.uk"
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
    - name: redpanda-console
      namespace: redpanda-ns
      port: 8080
---
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-cluster-issuer
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    email: artdevldn@gmail.com
    privateKeySecretRef:
      name: letsencrypt-cluster-issuer-key 
    solvers:
     - http01:
        ingress:
          ingressClassName: cilium 
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-app
spec:
  dnsNames:
    - artdevldn.co.uk  
  secretName: example-app-tls  
  issuerRef:
    name: letsencrypt-cluster-issuer
    kind: ClusterIssuer
```

Any idea how this could be fixed?
Thank you

**Comment by youngnick** [comment]
@arturs706, that error is a cert-manager error, saying that it could not reach through and perform the required HTTP01 check. I suggest two things, in order:
- review the cert-manager documentation and open a similar issue in the cert-manager repo if you can't find any help there
- open a separate issue in the Cilium repo and link this issue using the issue number, like #30720.

Commenting on unrelated closed issues is not the best way to get assistance.
